### DISCUSSION TITLE
How to use NextJS type URL template in SSR?

### DISCUSSION THREAD

**Discussion by muradsofi** [vercel/next.js#40790] [Help]
Let's say I have  URL templates like that 

1. `/properties/[id]/form` 
2. `/properties/[id]/preview`

And when I use this template like that it works perfectly 
 
router.push({
  pathname: "/properties/[id]/form",
  query: {
     id: 1
   }
});

But currently, I have one case in which I want to use this template in SSR (Inside getServerSideProps). 

return {
   redirect: {
          permanent: false,
          destination: "/properties/[id]/form",
  },
   props: {},
};

But looks like this is not working, do U know any other solution to use the NextJS type URL template in SSR? 

I try to find some npm library to convert this kind of URL template to a normal URL but didn't find :( 

It would be perfect if Next had `generatePath` function like react-router. (https://v5.reactrouter.com/core/api/generatePath)

**Comment by icyJoseph** [comment]
Hi,

What would be the value of `id` in this case? You need to be able to specify it as you redirect. Once you have it, `/properties/${id}/form` should do the trick.

## EDIT

OP wanted something more along the lines of:

```ts
const asPath = generateRouteAsPath("/properties/[id]/form", {
  id: 1,
}); // /properties/1/form 
```

My take is to do as [`next-router-mock` does](https://github.com/scottrippey/next-router-mock/blob/fbe8950695369c1d46de6c3cc9202187ef3d0ac0/src/MemoryRouter.tsx#L11), trying to remove all dependencies. For example this [TypeScript playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAJwKYFs4DdUBVkCGMANjGAOYDKxBAzgBYAUADgVPQFyK1TJnkBKRAG8AUIgkpUUEMiSt2AOlRgAJrQDqMdowBEAel1CAZMcQL6i4ivLtEAPkQBGRAH5zbS7VIRUjAAwANIgAtE5CXBYA3KIAvqKioJCwCIjk0gBKcCBQqACCtAAKnozikhZgBOioXDx8FIFlEgCOIKjIAJ5cGagQcMiqADx1-MEjFIgAPohgIOgARu1T3Lz8ANoAusuzC+2b9o2SiPR0nCv15NsgxMTL4KqowGSoqqJCYkd9YDxS6ERg-AAim1OogALwiRCKaGtdodRCxGJNRD6fSIHrMGi+bjEEDkWjBAhqX7YVCIdgYRDAZBwdDk+hkgAGsM6jOR1igiDoxTsEIqVVQijQmIIvkY+gAOmthE5AgAmWKMRQAalcAg2MvlsX05GCjAAJEEcXjaqsKEIwY4PkdJDBgIhGN48YoeARkFBNNomLpoYojEInZcIYGhagRWKfdDdMFdEYkTbJF8flgCLiyRC0H8yECQR01oGNgBCGIJiQPDlkzP-HNw-O48gbEulu0OvLIQgdRQwWhtjuMFNpgTvZGlqQyOSIAdtRR-ZiMfsWxwqPoPACqGQAkgBhWnMBAqKALgSKABWcDIekMAibCfio7Q46QU7JhbBEPuj2eqjciGXcDXm47uge5gAejCUGa5D9qmbRDogXCxjeCLXgkRyouiGCklyYDwrwRCkBM3inKgtBRPS3aIBARKIHAEAQLIiAthSaCMbQMxwIgfxQBA9BUv0XKUWwPEhKmtyBg6axrL6BYbAIyLcp44Ikjg+D4fw1CnIwCnsChyLoQAot8shkiy8LPmxbpkuM+LIkmnJoH0AxKQA8vMJ69FAyhgKsJGMFW2YUMCcLHmgqggL4gw9I5QzWWMkH2PY87MGgWDBGsADWqAdMEz6yeCVojuUKUZVlWwQmsuWnueYB6HGhUSA+sjyClSGxMEwixHJelonkzDMCo37+QCEymVyFmIKZsWMfaoEvC8HC2QgPzNGxEKgQA7og64ADKUKgbo8cUhDoLQfm9P0qjHlAcAQRcjC6UcLaMCtQjabxyoQoyrj6sIK2xIySFPScDCvUUikfccpzxpIjUTm9MTxKISZwNYVhwFB6RQFkOT5GDOiMvowBwHA+hrPMbobIT-ToIy7WIOTyAIU4ugIjG-oxMjqPEOjjCY9juQFDyTAE0TJNkxTpMwKolNE8gNN0wzTPRoxqhcE4-j+G1iCxkOHNLSjgrcxjmTZALePC1TYsM5TUnQgWtOQor2vM2M9ZcGsujq-4yuexr3tbFrOsoZzhs83zpu40LjAi8TpPW-o2gYA7wj026SuuyaiAe17Pte-7rPa+zohAA) shows a version without 3rd party dependencies.

I am not sure about the maintainability of this approach, **and I wouldn't recommend it** either. The GSSP layer should work independently of these constructs, it is only because OP has set out to build a map of paths ad-hoc, that this became necessary. We should avoid this kind of patterns, IMHO.

**Reply by muradsofi** [reply]
@icyJoseph Yes because of that I added an example react router's generatePath function. 

```
const URL = generatePath("/properties/[id]/form", {
  id: 1,
}); 
```

URL => /properties/1/form


But I don't find a similar implementation in NextJS. Problem is in getServerSideProps I already have `id` but I can't convert the dynamic URL template (`/properties/[id]/form`) to a normal URL (`/properties/1/form`)

**Reply by icyJoseph** [reply]
I must be missing something here. If you have the id value why not just use JavaScript string concatenation or template literals? 

```js
const id = 1
const destination = "/properties/"+ id + "/form"
// Or
const destination = `/properties/${id}/form`
```
 Then because it is a relative path, the request arrives at the file routing system and it is properly matched against it. 


That's all you need from GSSP. Am I missing something here? Perhaps I misunderstand your problem.

**Reply by icyJoseph** [reply]
Though even an absolute path works. 

Perhaps you are mixing how the client side router can make use of actual URL paths and those that reflect the file system structure, aka, `/post/9` vs `/post/[id]`

**Reply by muradsofi** [reply]
Yeah, it's a solution but not perfect. Why? 

I store all my page URLs as a template in one JSON file. 

const CLIENT_PATH = { 
  form: "/properties/[id]/form", 
  preview: "/properties/[id]/preview", 
  submission: "/properties/[id]/submission/"
}

And it's easy to convert URL template to URL with such a function `generatePath` instead of playing with string. 


If I play string for generating URL it can be a problem in the future. Let's say my URL changed this is mean I need to find these strings and replace them. But in my example, I need just to update the string in my CLIENT_PATH object.

**Reply by icyJoseph** [reply]
Aha, you see this is why we ought to share more of what we are trying to do.

I recall `next-router-mock` had run into a similar kind of situation, because they want to accomplish the same while testing. We can take a look at how they solved it here: 

https://github.com/scottrippey/next-router-mock/blob/fbe8950695369c1d46de6c3cc9202187ef3d0ac0/src/MemoryRouter.tsx#L11

The key pieces are:

```ts
/**
 * Creates a URL from a pathname + query.
 * Injects query params into the URL slugs, the same way that next/router does.
 */
function getRouteAsPath(pathname: string, query: ParsedUrlQuery, hash: string | null | undefined) {
```

And that you need to also get a hold of the `removeTrailingSlash` function.

You can play with a tweaked implementation of mine, that has no dependencies, [here](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAJwKYFs4DdUBVkCGMANjGAOYDKxBAzgBYAUADgVPQFyK1TJnkBKRAG8AUIgkpUUEMiSt2AOlRgAJrQDqMdowBEAel1CAZMcQL6i4ivLtEAPkQBGRAH5zbS7VIRUjAAwANIgAtE5CXBYA3KIAvqKioJCwCIjk0gBKcCBQqACCtAAKnozikhZgBOioXDx8FIFlEgCOIKjIAJ5cGagQcMiqADx1-MEjFIgAPohgIOgARu1T3Lz8ANoAusuzC+2b9o2SiPR0nCv15NsgxMTL4KqowGSoqqJCYkd9YDxS6ERg-AAim1OogALwiRCKaGtdodRCxGJNRD6fSIHrMGi+bjEEDkWjBAhqX7YVCIdgYRDAZBwdDk+hkgAGsM6jOR1igiDoxTsEIqVVQijQmIIvkY+gAOmthE5AgAmWKMRQAalcAg2MvlsX05GCjAAJEEcXjaqsKEIwY4PkdJDBgIhGN48YoeARkFBNNomLpoYojEInZcIYGhagRWKfdDdMFdEYkTbJF8flgCLiyRC0H8yECQR01oGNgBCGIJiQPDlkzP-HNw-O48gbEulu0OvLIQgdRQwWhtjuMFNpgTvZGlqQyOSIAdtRR-ZiMfsWxwqPoPACqGQAkgBhWnMBAqKALgSKABWcDIekMAibCfio7Q46QU7JhbBEPuj2eqjciGXcDXm47uge5gAejCUGa5D9qmbRDogXCxjeCLXgkRyouiGCklyYDwrwRCkBM3inKgtBRPS3aIBARKIHAEAQLIiAthSaCMbQMxwIgfxQBA9BUv0XKUWwPEhKmtyBg6axrL6BYbAIyLcp44Ikjg+D4fw1CnIwCnsChyLoQAot8shkiy8LPmxbpkuM+LIkmnJoH0AxKQA8vMJ69FAyhgKsJGMFW2YUMCcLHmgqggL4gw9I5QzWWMkH2PY87MGgWDBGsADWqAdMEz6yeCVojuUKUZVlWwQmsuWnueYB6HGhUSA+sjyClSGxMEwixHJelonkzDMCo37+QCEymVyFmIKZsWMfaoEvC8HC2QgPzNGxEKgQA7og64ADKUKgbo8cUhDoLQfm9P0qjHlAcAQRcjC6UcLaMCtQjabxyoQoyrj6sIK2xIySFPScDCvUUikfccpzxpIjUTm9MTxKISZwNYVhwFB6RQFkOT5GDOiMvowBwHA+hrPMbobIT-ToIy7WIOTyAIU4ugIjG-oxMjqPEOjjCY9juQFDyTAE0TJNkxTpMwKolNE8gNN0wzTPRoxqhcE4-j+G1iCxkOHNLSjgrcxjmTZALePC1TYsM5TUnQgWtOQor2vM2M9ZcGsujq-4yuexr3tbFrOsoZzhs83zpu40LjAi8TpPW-o2gYA7wj026SuuyaiAe17Pte-7rPa+zohAA).

We can see, that these inputs:

```ts
console.log(getRouteAsPath(`/foo/[bar]/form`, { bar: "1" }, ""));
console.log(getRouteAsPath(`/foo/[bar]/[id]/form`, { bar: "1", id: 100}, ""));
console.log(getRouteAsPath(`/foo/[bar]/[...slug]`, { bar: "1", slug: ["100", "1000"] }, ""));
console.log(getRouteAsPath(`/foo/[bar]/item`, { bar: "1", slug: ["100", "1000"] }, ""));
```

Log:


```
[LOG]: "/foo/1/form" 
[LOG]: "/foo/1/100/form" 
[LOG]: "/foo/1/100/1000" 
[LOG]: "/foo/1/item?slug=100%2C1000" 
```

**Reply by icyJoseph** [reply]
- @muradsofi did you try out the function I posted above?
